database/download.go

```diff
@@ -50,25 +60,33 @@ func (db *DB) DownloadByID(ctx context.Context, id primitive.ObjectID) (*Downloa
 
 // DownloadCreate registers a new download. Marks partial downloads by supplying
 // the `bytes` param. If `bytes` is 0 we assume a full download.
-func (db *DB) DownloadCreate(ctx context.Context, user User, skylink Skylink, bytes int64) (*Download, error) {
+func (db *DB) DownloadCreate(ctx context.Context, user User, skylink Skylink, bytes int64) error {
 	if user.ID.IsZero() {
-		return nil, errors.New("invalid user")
+		return errors.New("invalid user")
 	}
 	if skylink.ID.IsZero() {
-		return nil, errors.New("invalid skylink")
+		return errors.New("invalid skylink")
+	}
+
+	// Check if there exists a download of this skylink by this user, updated
+	// within the DownloadUpdateWindow and keep updating that, if so.
+	down, err := db.DownloadRecent(ctx, skylink.ID)
+	if err == nil && down.Updated.Add(DownloadUpdateWindow).After(time.Now().UTC()) {
+		// We found a recent download of this skylink. Let's update it.
+		return db.DownloadIncrement(ctx, down, bytes)
 	}
-	up := Download{
+
+	// We couldn't find a recent download of this skylink, updated within
```

Comment:
Should this be atomic? I think you'll quickly run into issues here where two concurrent calls to this endpoint wreak havoc in your database? Erroring out one of the two calls and such most probably if your database has any unqiue constraints.

We have unique constraints on all of these collections right? E.g. in downloads an entry should be unique on user and skylink for example.

Reply:
I'm not sure I'm following.

There is no unique constraint on user_id+skylink_id, so if we add a new download at this point it will get its own download_id with the same user_id and skylink_id as another one, if added in-between these calls.

In the race case where the browser sends two GETs very close together and:
1. we check the DB on the first GET, we see no record, we start creating it...
2. we check the DB on the second GET, we see no record (the first one is still pending creation), we start creating it...

We'll end up with two separate download records. This will pollute the user's download history slightly but it won't affect them in any other way because each created download will have its real size (as reported by nginx), so the user won't be charged for bandwidth they didn't consume.